_posts/2019-05-28-improved-hive-bucketing.md

```diff
@@ -0,0 +1,132 @@
+---
+layout: post
+title:  "Improved Hive Bucketing"
+author: David Phillips
+---
+
+The next release of Presto (version 312) adds support for the more
+flexible bucketing introduced in recent versions of Hive. Specifically,
+it allows any number of files per bucket, including zero. This allows
+inserting data into an existing partition without requiring a rewrite
+of the partition, and improves the performance of writes by no requiring
+the creation of files for empty buckets.
+
+# Hive bucketing overview
+
+Hive bucketing is a simple form of hash partitioning. A table is bucketed
+on one or more columns with a fixed number of hash buckets:
+
+```sql
+CREATE TABLE page_views (
+  user_id bigint,
+  page_url varchar,
+  dt date
+)
+WITH (
+  partitioned_by = ARRAY['dt'],
+  bucketed_by = ARRAY['user_id'],
+  bucket_count = 50
+)
+```
+
+The bucketing happens within each partition of the table. In the above
+example, the table is partitioned by date and is declared to have `50` buckets
+using the user ID column. This means that the table will have `50` buckets
+for each date. The assigned bucket for each row is determined by hashing
+the user ID value.
+
+# Original Hive bucketing
+
+Originally, Hive required exactly one file per bucket. The files were named 
+such that the bucket number was implicit based on the file's position within
+the lexicographic ordering of the file names. For example, the following list
+of files represent buckets `0` to `2`, respectively:
+
+```
+00000_0
+00001_0
+00002_0
+```
+
+```
+file0
+file3
+file5
+```
+
+```
+bucketA
+bucketB
+bucketD
+```
+
+The file names are meaningless aside from their ordering with respect to the
+other file names.
+
+# What's the problem?
+
+The original Hive bucketing scheme has a couple of problems:
+
+- Inserting data into the table by adding additional files is not possible.
+  Instead, an insert option requires rewriting all of the existing files,
```

Comment:
maybe drop the word "option"

Reply:
Oops, this was supposed to be “operation”